### DISCUSSION TITLE
State of the art/science of start-to-finish PCB design in Python

### DISCUSSION THREAD

**Discussion by r4dr3fr4d** [xesscorp/skidl#129] [Ideas]
As suggested by @devbisme, [in this issue](https://github.com/xesscorp/skidl/issues/127).  

SKIDL seems to satisfy the first part of Python-driven PCB design, replacing drawing a schematic in a GUI with describing it using Python in order to generate the netlist.  I'd like to gauge community interest in furthering this idea to designing and generating the PCB layout and the rest of the process, and enumerating what possibilities there are in existing tools.  

I'm pretty new to PCB design and may be pretty naive about this, so I'm also happy to take criticism on why this approach may not be feasible, or what roadblocks exist.  Hopefully any replies will help inform others who have the same question.

Thanks all!

**Comment by devbisme** [comment]
People who are well-versed in electronic circuit design are loath to use some automated tools (like autorouters or autoplacers) because they consider the results inferior to what they produce manually. This is probably justified for complex boards with many constraints.

But, an automated system for simple boards may be feasible. We had a discussion about this [here](https://github.com/michaelgale/pcbflow/issues/1#issuecomment-814981563) and [here](https://github.com/zignig/artifex/issues/1). Unfortunately, I don't think we achieved escape velocity.

However, I *did* start the [**Circuits as Code** project](https://github.com/devbisme/circuitsascode) to create a SKiDL library of common circuits. These could be used in an automated system.

There was some work at Berkeley on [JITPCB](https://people.eecs.berkeley.edu/~biancolin/papers/JITPCB.pdf), which was a tool for rapidly building the hardware and software for systems that were composed of subsystems that communicated over a narrow bus. I think some of the people who worked on that went off and founded [jitx](https://www.jitx.com/) as a commercial offshoot. They also have a [VSCode extension](https://marketplace.visualstudio.com/items?itemName=JITX.jitpcb-vscode).

**Comment by r4dr3fr4d** [comment]
I'm honestly not even interested in automation.  For example, I'm totally
fine with describing tracks (correct term?) as, say segments defined using
some Python object.  I just want to keep the GUI out of the picture.

Does that address your point or clarify things?

**Comment by devbisme** [comment]
Something that's kinda similar to what you want is James Bowman's [CuFlow](https://github.com/jamesbowman/cuflow). As he [describes](https://excamera.com/sphinx/article-cuflow-crossbar.html) it: "Instead of spending hours working away at a PCB CAD system, you can spend hours on a Python program that draws the PCB instead!"

Another way to go is just to enter PCBNEW's Python console and use the `pcbnew` module to place components and create traces. You could captured the commands and place them into a Python script that would re-create the board whenever you ran it.

Once the layout is done, there are some existing [script-based flows](https://github.com/devbisme/kicad-3rd-party-tools#manufacturing-output-tools) for generating the manufacturing outputs.

Let's keep this conversation going. Let me know if there are things I can do in SKiDL to help your efforts.

**Reply by r4dr3fr4d** [reply]
@devbisme  I did take a look at CuFlow and since have discovered this: https://github.com/michaelgale/pcbflow 

It seems to be a derivative and has a basic example on the README.  I will try to see what I can do with it - I'm trying to make a simple board that will connect a single Cherry MX switch to a row/column switch matrix.  It'll look like the following model:

![cherry_mx_1u_pcb](https://user-images.githubusercontent.com/10680546/135857544-bbb5d2d0-cb1b-447b-ba42-7c014bf958cc.png)

**Reply by devbisme** [reply]
pcbflow looks interesting!
